File: app/schemas/formcreator.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel

class ServiceCategory(BaseModel):
    id: int
    name: str
    parent_id: int
    level: int
    completename: Optional[str] = None

class ServiceForm(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    category_id: int
    icon: Optional[str] = None
    icon_color: Optional[str] = None
    background_color: Optional[str] = None

class FormLookupRef(BaseModel):
    source: str
    params: Dict[str, Any] = {}

class FormOption(BaseModel):
    label: str
    value: Any

class FormQuestion(BaseModel):
    id: int
    name: str
    fieldtype: str
    required: bool
    description: Optional[str] = None
    default_value: Any = None
    options: Optional[List[FormOption]] = None
    lookup: Optional[FormLookupRef] = None
    layout: Dict[str, int] = {}
    show_rule: Optional[int] = None

class FormSection(BaseModel):
    id: int
    name: str
    order: int
    questions: List[FormQuestion]
    show_rule: Optional[int] = None

class FormCondition(BaseModel):
    id: int
    controller_question_id: int
    target_itemtype: str
    target_items_id: int
    show_condition: int
    show_logic: int
    show_value: str
    order: int

class FormSchema(BaseModel):
    form: Dict[str, Any]
    sections: List[FormSection]
    conditions: List[FormCondition]
    regexes: List[Dict[str, Any]] = []
    ranges: List[Dict[str, Any]] = []

class SubmitFormRequest(BaseModel):
    answers: Dict[str, Any]

class SubmitFormResponse(BaseModel):
    form_answer_id: int
    message: str
    ticket_ids: List[int] = []

class LookupItem(BaseModel):
    id: int
    name: str
    completename: Optional[str] = None

class LookupResponse(BaseModel):
    items: List[LookupItem]

File: web/src/lib/api/contracts/formcreator.ts
export interface ServiceCategoryDto {
  id: number;
  name: string;
  parent_id: number;
  level: number;
  completename?: string;
}

export interface ServiceFormDto {
  id: number;
  name: string;
  description?: string;
  category_id: number;
  icon?: string;
  icon_color?: string;
  background_color?: string;
}

export interface FormOptionDto {
  label: string;
  value: string | number;
}

export interface FormLookupRefDto {
  source: string;
  params: Record<string, unknown>;
}

export interface ApiFormQuestionDto {
  id: number;
  name: string;
  fieldtype: string;
  required: boolean;
  description?: string;
  default_value?: unknown;
  options?: FormOptionDto[];
  lookup?: FormLookupRefDto;
  layout: { row: number; col: number; width: number };
  show_rule?: number;
}

export interface ApiFormSectionDto {
  id: number;
  name: string;
  order: number;
  questions: ApiFormQuestionDto[];
  show_rule?: number;
}

export interface ApiFormConditionDto {
  id: number;
  controller_question_id: number;
  target_itemtype: string;
  target_items_id: number;
  show_condition: number;
  show_logic: number;
  show_value: string;
  order: number;
}

export interface ApiFormSchemaDto {
  form: Record<string, unknown>;
  sections: ApiFormSectionDto[];
  conditions: ApiFormConditionDto[];
  regexes: Record<string, unknown>[];
  ranges: Record<string, unknown>[];
}

export interface SubmitFormResponseDto {
  form_answer_id: number;
  message: string;
  ticket_ids: number[];
}
